In [1]:
cd("../../../../temp")

In [2]:
]activate temp

  Activating project at `c:\Users\Work\Documents\Personal\Work\classes\PUCP 2024-II\repo\temp\temp`


In [3]:
cd("../Labs/Assignment/Assignment_1/Solutions")

In [4]:
using DataFrames
using Statistics
using StatsBase
using CSV
using StatsModels
using GLM

# 1 Data analysis

1. Import the data set. Make sure the column names are imported as intended.

In [5]:
path = "C:/Users/User/Documents/GitHub/CausalAI-Course-1/"
data = CSV.read("../../../../data/wage2015_subsample_inference.csv", DataFrame) 
select!(data, Not("rownames"))

5150×20 DataFrame
  Row │ wage      lwage    sex      shs      hsg      scl      clg      ad     ⋯
      │ Float64   Float64  Float64  Float64  Float64  Float64  Float64  Float6 ⋯
──────┼─────────────────────────────────────────────────────────────────────────
    1 │  9.61538  2.26336      1.0      0.0      0.0      0.0      1.0      0. ⋯
    2 │ 48.0769   3.8728       0.0      0.0      0.0      0.0      1.0      0.
    3 │ 11.0577   2.40313      0.0      0.0      1.0      0.0      0.0      0.
    4 │ 13.9423   2.63493      1.0      0.0      0.0      0.0      0.0      1.
    5 │ 28.8462   3.36198      1.0      0.0      0.0      0.0      1.0      0. ⋯
    6 │ 11.7308   2.46222      1.0      0.0      0.0      0.0      1.0      0.
    7 │ 19.2308   2.95651      1.0      0.0      1.0      0.0      0.0      0.
    8 │ 19.2308   2.95651      0.0      0.0      1.0      0.0      0.0      0.
    9 │ 12.0      2.48491      1.0      0.0      1.0      0.0      0.0      0. ⋯
   10 │ 19.2308   2.95651      1.0      0.0      0.0      0.0      1.0      0.
   11 │ 17.3077   2.85115      1.0      0.0      1.0      0.0      0.0      0.
  ⋮   │    ⋮         ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮   ⋱
 5141 │ 45.5466   3.81874      1.0      0.0      0.0      0.0      1.0      0.
 5142 │ 22.5962   3.11778      0.0      0.0      1.0      0.0      0.0      0. ⋯
 5143 │ 16.8269   2.82298      0.0      0.0      1.0      0.0      0.0      0.
 5144 │ 24.0385   3.17966      1.0      0.0      0.0      0.0      1.0      0.
 5145 │ 13.8462   2.62801      0.0      0.0      0.0      0.0      1.0      0.
 5146 │ 14.7692   2.69255      0.0      0.0      0.0      0.0      1.0      0. ⋯
 5147 │ 23.0769   3.13883      1.0      0.0      0.0      1.0      0.0      0.
 5148 │ 38.4615   3.64966      0.0      0.0      0.0      0.0      0.0      1.
 5149 │ 32.967    3.49551      0.0      0.0      1.0      0.0      0.0      0.
 5150 │ 17.3077   2.85115      0.0      0.0      0.0      0.0      0.0      1. ⋯
                                                13 columns and 5129 rows omitted

2. Are there missing values? Display the number of missings (if any) of each variable in the data set.

In [6]:
describe(data, :nmissing)

20×2 DataFrame
 Row │ variable  nmissing 
     │ Symbol    Int64    
─────┼────────────────────
   1 │ wage             0
   2 │ lwage            0
   3 │ sex              0
   4 │ shs              0
   5 │ hsg              0
   6 │ scl              0
   7 │ clg              0
   8 │ ad               0
   9 │ mw               0
  10 │ so               0
  11 │ we               0
  12 │ ne               0
  13 │ exp1             0
  14 │ exp2             0
  15 │ exp3             0
  16 │ exp4             0
  17 │ occ              0
  18 │ occ2             0
  19 │ ind              0
  20 │ ind2             0

3. Report descriptive statistics of the variables (mean, standard deviation, percetiles, etc.). Interpret your results.

In [7]:
describe(data, :min, :max, :mean, :std, :q25, :median, :q75)

20×8 DataFrame
 Row │ variable  min        max          mean         std           q25        ⋯
     │ Symbol    Real       Real         Float64      Float64       Float64    ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ wage        3.02198     528.846     23.4104       21.003       13.4615  ⋯
   2 │ lwage       1.10591       6.2707     2.97079       0.570385     2.59984
   3 │ sex         0.0           1.0        0.444466      0.496955     0.0
   4 │ shs         0.0           1.0        0.023301      0.150872     0.0
   5 │ hsg         0.0           1.0        0.243883      0.429465     0.0     ⋯
   6 │ scl         0.0           1.0        0.278058      0.448086     0.0
   7 │ clg         0.0           1.0        0.31767       0.465616     0.0
   8 │ ad          0.0           1.0        0.137087      0.343973     0.0
   9 │ mw          0.0           1.0        0.259612      0.438464     0.0     ⋯
  10 │ so          0.0           1.0        0.296505      0.456761     0.0
  11 │ we          0.0           1.0        0.216117      0.411635     0.0
  12 │ ne          0.0           1.0        0.227767      0.419432     0.0
  13 │ exp1        0.0          47.0       13.7606       10.6095       5.0     ⋯
  14 │ exp2        0.0          22.09       3.01893       4.0009       0.25
  15 │ exp3        0.0         103.823      8.23587      14.489        0.125
  16 │ exp4        0.0         487.968     25.118        53.5302       0.0625
  17 │ occ        10.0      100000.0     5310.74      11874.4       1740.0     ⋯
  18 │ occ2        1            22         11.6709        6.96668      5.0
  19 │ ind       370.0      100000.0     6629.15       5333.44      4880.0
  20 │ ind2        2            22         13.3169        5.70102      9.0
                                                               2 columns omitted

As most of the variables are dummy variables, only wage is interprated based in these descriptive statistics. The average wage in the sample is 23.410 dollars per hour. The 25% richest earn 27.78 dollars per hour or more, while the 25% poorest earn 13.45 or less. The median earning per hour is 19.23 dollars, which is lower than the average wage and shows the asymmetry in the wage distribution of the sample. 

4. How many women with a college graduate degree (clg) or above have a wage corresponding to the 25% richest of the sample? Report the dataframe corresponding to this conditions and the result.

In [8]:
data_women_clg_75 = data[(data.sex .== 1) .&
                         ((data.clg .== 1) .| (data.ad .== 1)) .&
                         (data.wage.>describe(data, :q75)[:,2][1]),:]
data_women_clg_75

412×20 DataFrame
 Row │ wage      lwage    sex      shs      hsg      scl      clg      ad      ⋯
     │ Float64   Float64  Float64  Float64  Float64  Float64  Float64  Float64 ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │  28.8462  3.36198      1.0      0.0      0.0      0.0      1.0      0.0 ⋯
   2 │  42.3077  3.74497      1.0      0.0      0.0      0.0      0.0      1.0
   3 │  41.2088  3.71865      1.0      0.0      0.0      0.0      1.0      0.0
   4 │ 100.0     4.60517      1.0      0.0      0.0      0.0      0.0      1.0
   5 │  33.6538  3.51613      1.0      0.0      0.0      0.0      1.0      0.0 ⋯
   6 │  36.0577  3.58512      1.0      0.0      0.0      0.0      1.0      0.0
   7 │  31.25    3.44202      1.0      0.0      0.0      0.0      1.0      0.0
   8 │  28.8462  3.36198      1.0      0.0      0.0      0.0      1.0      0.0
   9 │  48.0769  3.8728       1.0      0.0      0.0      0.0      1.0      0.0 ⋯
  10 │  34.6154  3.5443       1.0      0.0      0.0      0.0      1.0      0.0
  11 │  31.25    3.44202      1.0      0.0      0.0      0.0      0.0      1.0
  ⋮  │    ⋮         ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮    ⋱
 403 │  40.3846  3.69845      1.0      0.0      0.0      0.0      1.0      0.0
 404 │  65.3846  4.18029      1.0      0.0      0.0      0.0      0.0      1.0 ⋯
 405 │  29.4872  3.38396      1.0      0.0      0.0      0.0      1.0      0.0
 406 │  28.8462  3.36198      1.0      0.0      0.0      0.0      0.0      1.0
 407 │  50.3635  3.91927      1.0      0.0      0.0      0.0      0.0      1.0
 408 │  48.0769  3.8728       1.0      0.0      0.0      0.0      0.0      1.0 ⋯
 409 │  30.1455  3.40604      1.0      0.0      0.0      0.0      1.0      0.0
 410 │  29.9145  3.39834      1.0      0.0      0.0      0.0      1.0      0.0
 411 │  48.0769  3.8728       1.0      0.0      0.0      0.0      1.0      0.0
 412 │  45.5466  3.81874      1.0      0.0      0.0      0.0      1.0      0.0 ⋯
                                                 12 columns and 391 rows omitted

In [9]:
num_women = size(data_women_clg_75)[1]
println("$num_women women with a college graduate degree (clg) or above have a wage corresponding to the 25% richest of the sample")

412 women with a college graduate degree (clg) or above have a wage corresponding to the 25% richest of the sample


5. How many men with a high school graduate degree (hsg) or below have a wage corresponding to the 25% richest of the sample? Report the dataframe corresponding to this conditions and the result.

In [10]:
data_men_hsg_75 = data[(data.sex .== 0) .&
                       ((data.shs .== 1) .| (data.hsg .== 1)) .&
                       (data.wage.>describe(data, :q75)[:,2][1]),:]
data_men_hsg_75

115×20 DataFrame
 Row │ wage      lwage    sex      shs      hsg      scl      clg      ad      ⋯
     │ Float64   Float64  Float64  Float64  Float64  Float64  Float64  Float64 ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │  27.8846  3.32808      0.0      0.0      1.0      0.0      0.0      0.0 ⋯
   2 │  28.8462  3.36198      0.0      0.0      1.0      0.0      0.0      0.0
   3 │  28.8462  3.36198      0.0      0.0      1.0      0.0      0.0      0.0
   4 │  28.8462  3.36198      0.0      0.0      1.0      0.0      0.0      0.0
   5 │  29.7143  3.39163      0.0      0.0      1.0      0.0      0.0      0.0 ⋯
   6 │  35.5769  3.5717       0.0      0.0      1.0      0.0      0.0      0.0
   7 │  36.3248  3.5925       0.0      0.0      1.0      0.0      0.0      0.0
   8 │  43.2692  3.76744      0.0      0.0      1.0      0.0      0.0      0.0
   9 │  43.2692  3.76744      0.0      0.0      1.0      0.0      0.0      0.0 ⋯
  10 │  43.2692  3.76744      0.0      0.0      1.0      0.0      0.0      0.0
  11 │  43.2692  3.76744      0.0      0.0      1.0      0.0      0.0      0.0
  ⋮  │    ⋮         ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮    ⋱
 106 │  38.4615  3.64966      0.0      0.0      1.0      0.0      0.0      0.0
 107 │  48.0774  3.87281      0.0      0.0      1.0      0.0      0.0      0.0 ⋯
 108 │ 115.385   4.74827      0.0      0.0      1.0      0.0      0.0      0.0
 109 │  33.6538  3.51613      0.0      0.0      1.0      0.0      0.0      0.0
 110 │  43.2692  3.76744      0.0      1.0      0.0      0.0      0.0      0.0
 111 │  33.6538  3.51613      0.0      0.0      1.0      0.0      0.0      0.0 ⋯
 112 │  40.8654  3.71028      0.0      0.0      1.0      0.0      0.0      0.0
 113 │  31.25    3.44202      0.0      0.0      1.0      0.0      0.0      0.0
 114 │  33.6538  3.51613      0.0      0.0      1.0      0.0      0.0      0.0
 115 │  32.967   3.49551      0.0      0.0      1.0      0.0      0.0      0.0 ⋯
                                                  12 columns and 94 rows omitted

In [11]:
num_men = size(data_men_hsg_75)[1]
println("$num_men men with a high school degree (hsg) or below have a wage corresponding to the 25% richest of the sample")

115 men with a high school degree (hsg) or below have a wage corresponding to the 25% richest of the sample


6. Create two dataframes. One containing only the log(wage) and the other containig every variable of the data set but the wage related variables.

In [12]:
log_wage = select(data, "lwage")
log_wage


5150×1 DataFrame
  Row │ lwage   
      │ Float64 
──────┼─────────
    1 │ 2.26336
    2 │ 3.8728
    3 │ 2.40313
    4 │ 2.63493
    5 │ 3.36198
    6 │ 2.46222
    7 │ 2.95651
    8 │ 2.95651
    9 │ 2.48491
   10 │ 2.95651
   11 │ 2.85115
  ⋮   │    ⋮
 5141 │ 3.81874
 5142 │ 3.11778
 5143 │ 2.82298
 5144 │ 3.17966
 5145 │ 2.62801
 5146 │ 2.69255
 5147 │ 3.13883
 5148 │ 3.64966
 5149 │ 3.49551
 5150 │ 2.85115
5129 rows omitted

In [13]:
features = select(data, Not("lwage","wage"))
features

5150×18 DataFrame
  Row │ sex      shs      hsg      scl      clg      ad       mw       so      ⋯
      │ Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64 ⋯
──────┼─────────────────────────────────────────────────────────────────────────
    1 │     1.0      0.0      0.0      0.0      1.0      0.0      0.0      0.0 ⋯
    2 │     0.0      0.0      0.0      0.0      1.0      0.0      0.0      0.0
    3 │     0.0      0.0      1.0      0.0      0.0      0.0      0.0      0.0
    4 │     1.0      0.0      0.0      0.0      0.0      1.0      0.0      0.0
    5 │     1.0      0.0      0.0      0.0      1.0      0.0      0.0      0.0 ⋯
    6 │     1.0      0.0      0.0      0.0      1.0      0.0      0.0      0.0
    7 │     1.0      0.0      1.0      0.0      0.0      0.0      0.0      0.0
    8 │     0.0      0.0      1.0      0.0      0.0      0.0      0.0      0.0
    9 │     1.0      0.0      1.0      0.0      0.0      0.0      0.0      0.0 ⋯
   10 │     1.0      0.0      0.0      0.0      1.0      0.0      0.0      0.0
   11 │     1.0      0.0      1.0      0.0      0.0      0.0      0.0      0.0
  ⋮   │    ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮    ⋱
 5141 │     1.0      0.0      0.0      0.0      1.0      0.0      0.0      0.0
 5142 │     0.0      0.0      1.0      0.0      0.0      0.0      0.0      0.0 ⋯
 5143 │     0.0      0.0      1.0      0.0      0.0      0.0      0.0      0.0
 5144 │     1.0      0.0      0.0      0.0      1.0      0.0      0.0      0.0
 5145 │     0.0      0.0      0.0      0.0      1.0      0.0      0.0      0.0
 5146 │     0.0      0.0      0.0      0.0      1.0      0.0      0.0      0.0 ⋯
 5147 │     1.0      0.0      0.0      1.0      0.0      0.0      0.0      0.0
 5148 │     0.0      0.0      0.0      0.0      0.0      1.0      0.0      0.0
 5149 │     0.0      0.0      1.0      0.0      0.0      0.0      0.0      0.0
 5150 │     0.0      0.0      0.0      0.0      0.0      1.0      0.0      0.0 ⋯
                                                10 columns and 5129 rows omitted

# 2. Data Wrangling

In [14]:
data = CSV.read(
        download("https://raw.githubusercontent.com/d2cml-ai/CausalAI-Course/main/data/wage2015_subsample_inference.csv"), 
        DataFrame, 
        types = Dict(:occ2 => String, :ind2 => String)
);

7.

In [15]:
y = data[:, "lwage"];

8. Now that we have seen the for loops are implemented in Python, they can easily be ported into Julia, so we will not be focusing on that here. However, remember that the formula string methods can cause some issues as they are abstractions of larger procedures "under the hood," meaning that we lose some control over exactly how the generation works.

8.1.

In [16]:
basic_formula = @formula(lwage ~ 1 + sex + hsg + scl + clg + ad + so + we + ne + exp1 + occ2 + ind2)
X_basic = modelmatrix(basic_formula, data);

8.2.

In [17]:
flexible_formula = @formula(lwage ~ 1 + sex + (exp1 + exp2 + exp3 + exp4) * (hsg + scl + clg + ad + so + we + ne + occ2 + ind2))
X_flexible = modelmatrix(flexible_formula, data);

8.3

In [18]:
extra_flexible_formula = @formula(lwage ~ 1 + sex + (exp1 + exp2 + exp3 + exp4) * (hsg + scl + clg + ad + so + we + ne + occ2 + ind2) + (hsg + scl + clg + ad) * (so + we + ne + occ2 + ind2) + (so + we + ne) * (occ2 + ind2) + occ2 * ind2)
X_extra_flexible = modelmatrix(extra_flexible_formula, data);

9.

In [19]:
train_sample = rand(Float64, size(data)[1]) .< 0.8
test_sample = .!(train_sample)
y_train, y_test = y[train_sample], y[test_sample]
X_basic_train, X_basic_test = X_basic[train_sample, :], X_basic[test_sample, :]
X_flexible_train, X_flexible_test = X_flexible[train_sample, :], X_flexible[test_sample, :]
X_extra_flexible_train, X_extra_flexible_test = X_extra_flexible[train_sample, :], X_extra_flexible[test_sample, :];

# 3. Estimation

10

In [20]:
basic_model = lm(X_basic_train, y_train);

In [21]:
flexible_model = lm(X_flexible_train, y_train);

In [22]:
extra_flexible_model = lm(X_extra_flexible_train, y_train);

11

In [23]:
basic_mse_training = mean(residuals(basic_model) .^ 2)
basic_r2_training = 1 - basic_mse_training / var(y_train)
basic_adjr2_training = 1 - size(y_train)[1] / (size(y_train)[1] - size(X_basic)[2]) * basic_mse_training / var(y_train)
basic_mse_testing = mean((predict(basic_model, X_basic_test) - y_test) .^ 2)
basic_r2_testing = 1 - basic_mse_testing / var(y_test)

println("The training MSE for the basic model is $basic_mse_training")
println("The training R2 for the basic model is $basic_r2_training")
println("The training Adjusted R2 for the basic model is $basic_adjr2_training")
println("The testing MSE for the basic model is $basic_mse_testing")
println("The testing R2 for the basic model is $basic_r2_testing")

The training MSE for the basic model is 0.22348590367332788
The training R2 for the basic model is 0.31678162069303994
The training Adjusted R2 for the basic model is 0.30828512292550647
The testing MSE for the basic model is 0.23239308168478387
The testing R2 for the basic model is 0.26986458958513915


In [24]:
flexible_mse_training = mean(residuals(flexible_model) .^ 2)
flexible_r2_training = 1 - flexible_mse_training / var(y_train)
flexible_adjr2_training = 1 - size(y_train)[1] / (size(y_train)[1] - size(X_flexible)[2]) * flexible_mse_training / var(y_train)
flexible_mse_testing = mean((predict(flexible_model, X_flexible_test) - y_test) .^ 2)
flexible_r2_testing = 1 - flexible_mse_testing / var(y_test)

println("The training MSE for the flexible model is $flexible_mse_training")
println("The training R2 for the flexible model is $flexible_r2_training")
println("The training Adjusted R2 for the flexible model is $flexible_adjr2_training")
println("The testing MSE for the flexible model is $flexible_mse_testing")
println("The testing R2 for the flexible model is $flexible_r2_testing")

The training MSE for the flexible model is 0.2085879983205243
The training R2 for the flexible model is 0.36232598202819155
The training Adjusted R2 for the flexible model is 0.3221652527857275
The testing MSE for the flexible model is 0.24573214602313082
The testing R2 for the flexible model is 0.22795575501647714


In [25]:
extra_flexible_mse_training = mean(residuals(extra_flexible_model) .^ 2)
extra_flexible_r2_training = 1 - extra_flexible_mse_training / var(y_train)
extra_flexible_adjr2_training = 1 - size(y_train)[1] / (size(y_train)[1] - size(X_extra_flexible)[2]) * extra_flexible_mse_training / var(y_train)
extra_flexible_mse_testing = mean((predict(extra_flexible_model, X_extra_flexible_test) - y_test) .^ 2)
extra_flexible_r2_testing = 1 - extra_flexible_mse_testing / var(y_test)

println("The training MSE for the extra_flexible model is $extra_flexible_mse_training")
println("The training R2 for the extra_flexible model is $extra_flexible_r2_training")
println("The training Adjusted R2 for the extra_flexible model is $extra_flexible_adjr2_training")
println("The testing MSE for the extra_flexible model is $extra_flexible_mse_testing")
println("The testing R2 for the extra_flexible model is $extra_flexible_r2_testing")

The training MSE for the extra_flexible model is 0.17242648368025523
The training R2 for the extra_flexible model is 0.47287528746413043
The training Adjusted R2 for the extra_flexible model is 0.3132658279105961
The testing MSE for the extra_flexible model is 0.2813066618551167
The testing R2 for the extra_flexible model is 0.11618730851630032
